packages/nimble-components/src/table/specs/README.md

```diff
@@ -132,13 +132,17 @@ The various APIs/features of the `nimble-table` will be split up amongst several
     -   Define how the action menu gets associated with a particular column
     -   Define the table-level(column-level?) API(s) for applying an action menu to the table (slot, properties, etc...)
 
+_Attributes_
+
+-   `id-field-name` - An optional string attribute that specifies the field name within a row's record to use as a row's ID. If the attribute is not defined or a field with the specified name does not exist in a row's record, a default ID will be created. The default ID is the index of the row's record within the table's `data` property. The value associated with the field name must be unique across all records in the table's data.
```

Comment:
Given the complexity of this attribute, I think we should have a readonly property (not attribute) that is an enum of the resolved id mode, maybe `idMode='field' | 'index'` or something. Seems like it would be useful for debugging.

Reply:
Sure, I can add that. Are you envisioning the logic being something as simple as:
```ts
this.idMode = this.idFieldName ? 'field' : 'index';
```

Or, are you picturing different logic?

Reply:
From the description it sounds like there is more to the logic

> If the attribute is not defined or a field with the specified name does not exist in a row's record, a default ID will be created.

```ts
this.idMode = isInRecords(this.idFieldName) ? 'field' : 'index';
```

Reply:
Maybe I was misunderstanding some of the other feedback, but it sounded to me like part of the validation logic we wanted to add was to verify that the field name was valid for the records. Therefore, I was thinking we wouldn't have a valid case where the field wasn't in the records.

Reply:
Yea if we do strict validation that makes sense. Maybe we need an 'error' case?

```ts
const table = document.createElement('nimble-table');
table.idFieldName= 'test';
table.data = // some invalid value with wrong id. Is the assignment ignored altogether (keep default / old value)? I think... maybe.. we take the data but don't render the table with it. The user could run these assignments in different orders, ie. assign data then idFieldName vs idFoeldName then table
console.log(table.idMode) // 'error' ?
```

```ts
const table = document.createElement('nimble-table');
table.idFieldName= 'test';
table.data = // assign some good data, [{test: 1, other: 'hello'}, {test: 2, other: 'world'}]
console.log(table.idMode) // field
table.idFieldName= 'bob';
console.log(table.idMode) // with strict validation I guess this would now be 'error'?
```

Reply:
The question in my mind is what happens when the validation detects an issue? Does it throw an exception, throw out the data and not apply it to the table, something else? I ask because I would think the purpose of validation is keeping someone from getting into a case that is invalid.

I'm also having trouble understanding the purpose of this property -- maybe because of the questions I just posed. It seems to me that a client would know how they've configured the table (either to use field-based IDs or default IDs), so I'm not sure how this property would actually be used (other than perhaps by our own unit tests).

Reply:
> I'm also having trouble understanding the purpose of this property 

The initial spec text where this property was suggested included conditional logic to resolve the expected behavior before we discussed strict validation. In the initial text the table silently ignored an invalid idFieldName and switched to index mode internally without telling the user.

The property was to make that hidden choice available to users to understand the state of the table. Primarily for user debugging. Depending on the direction we go with strict validation it may not be necessary. My end goal is that if the table is ignoring behaviors / making choices, that the user can observe the resolved state of the table's choices.

> The question in my mind is what happens when the validation detects an issue? Does it throw an exception, throw out the data and not apply it to the table, something else?

Yep that is the question. If setting the data was done by a method then throwing an exception makes sense. When setting the data by a property I don't think that is as good an idea to throw an error from the accessor...

One concern I have with throwing an exception is how do the two properties relate. The problem is one effects the other and potentially invalidates the other. We should not enforce an order requirement in the assignment.

Reply:
Maybe we can take a page out of the HTML 5 form validation api and have a [validity state](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState/badInput) object?

```ts
const table = document.createElement('nimble-table');
table.checkValidity() // true, empty data is valid data
table.data = [{a:0},{a:1}]
table.checkValidity() // true
table.data = [{a:0},{a:1, id: 1}]
table.checkValidity() // true
table.idFieldName = 'invalid';
table.checkValidity() // false
table.validity.invalidIdField === true;
table.idFieldName = 'id';
table.checkValidity() // false
table.validity.recordsMissingIdField === true;
```

The assumpton being that we take the properties assigned, don't reject or throw. And if the properties are invalid we ignore them / use the default of an empty table.

This pattern will probably be useful for column configuration as well, i.e. the field name for a column configuration is not in the data, etc.
```

Reply:
I've updated the spec as follows:
- we'll always do validation
- if validation fails, no rows will be rendered in the table
- the table will report that it is invalid through a `validity` property and a `checkValidity()` function

Given this, I don't think we need an `idMode` property since a table with valid configuration will always use the field specified by `id-field-name`, unless that attribute is undefined.